server/datastore/mysql/software.go

```diff
@@ -211,29 +211,56 @@ func (d *Datastore) hostSoftwareFromHostID(tx *sqlx.Tx, id uint) ([]fleet.Softwa
 		selectFunc = tx.Select
 	}
 	sql := `
-		SELECT s.id, s.name, s.version, s.source, coalesce(scp.cpe, "") as generated_cpe,
-			IF(
-				COUNT(scv.cve) = 0,
-				null,
-				GROUP_CONCAT(
-					JSON_OBJECT(
-						"cve", scv.cve,
-						"details_link", CONCAT('https://nvd.nist.gov/vuln/detail/', scv.cve)
-					)
-				)
-			) as vulnerabilities FROM software s
+		SELECT s.id, s.name, s.version, s.source, coalesce(scp.cpe, "") as generated_cpe
+		FROM software s
 		LEFT JOIN software_cpe scp ON (s.id=scp.software_id)
-		LEFT JOIN software_cve scv ON (scp.id=scv.cpe_id)
 		WHERE s.id IN
 			(SELECT software_id FROM host_software WHERE host_id = ?)
 		group by s.id, s.name, s.version, s.source, generated_cpe
 	`
-	var result []fleet.Software
+	var result []*fleet.Software
 	if err := selectFunc(&result, sql, id); err != nil {
 		return nil, errors.Wrap(err, "load host software")
 	}
 
-	return result, nil
+	sql = `
+		SELECT s.id, scv.cve
+		FROM software s
+		JOIN software_cpe scp ON (s.id=scp.software_id)
+		JOIN software_cve scv ON (scp.id=scv.cpe_id)
+		WHERE s.id IN
+			(SELECT software_id FROM host_software WHERE host_id = ?)
+	`
+	queryFunc := d.db.Queryx
+	if tx != nil {
+		queryFunc = tx.Queryx
+	}
+
+	rows, err := queryFunc(sql, id)
+	if err != nil {
+		return nil, errors.Wrap(err, "load host software")
+	}
+	defer rows.Close()
+
+	cvesBySoftware := make(map[uint]fleet.VulnerabilitiesSlice)
+	for rows.Next() {
+		var id uint
+		var cve string
+		if err := rows.Scan(&id, &cve); err != nil {
+			return nil, errors.Wrap(err, "scanning cve")
+		}
+		cvesBySoftware[id] = append(cvesBySoftware[id], fleet.SoftwareCVE{
+			CVE:         cve,
+			DetailsLink: fmt.Sprintf("https://nvd.nist.gov/vuln/detail/%s", cve),
+		})
+	}
```

Comment:
There should be a check for any iteration error here, after the loop (the static golangci-lint should catch it):

```
if err := rows.Err(); err != nil {
  // ... handle error
}
```